src/Streamly/Internal/Data/Stream/StreamD/Type.hs

```diff
@@ -583,54 +586,86 @@ take n (Stream step state) = n `seq` Stream step' (state, 0)
 ------------------------------------------------------------------------------
 
 -- s = stream state, fs = fold state
-data GroupState s fs
+{-# ANN type GroupState Fuse #-}
+data GroupState s fs b a
     = GroupStart s
-    | GroupBuffer s fs Int
-    | GroupYield fs (GroupState s fs)
+    | GroupConsume s fs a
```

Comment:
`GroupConsume` is technically not needed but looks like the performance is affected without it. I did not check it with the plugin. But we can eliminate this later. I'd not make any more changes in this PR now.